pkg/app/api/service/pipedservice/service.proto

```diff
@@ -477,3 +481,16 @@ message ReportUnregisteredApplicationConfigurationsRequest {
 
 message ReportUnregisteredApplicationConfigurationsResponse {
 }
+
+message CreateDeploymentChainRequest {
+    message ApplicationsFilter {
+        string app_name = 1;
+        string app_kind = 2;
+        map<string,string> app_labels = 3;
+    }
+
```

Comment:
How about using the same `Node` as it is?

Reply:
👍

Reply:
tbh, I'm just want to rename the `Node` to `ApplicationsFilter` since those things are from the `spec.postSync.chain` configuration and at the time we load it from the configuration file, it's kind of filter used to find applications more than `node` or `applications`, the `node` word come when we create deployment chain, and `applications` come from users side, just to notify users that the filters written there used to find out applications for the chain.

Reply:
Ok. I get your points.

So it is time to find some good names for all those things: name in the config, name in gRPC request, name in chain model. 🤔 

How do you think about this?


``` go

// config

type DeploymentChain struct {
	Applications []ChainApplicationMatcher`json:"applications"`
	// Conditions *ChainTriggerCondition `json:"conditions,omitempty"`
}

type ChainApplicationMatcher struct {
	Name   string            `json:"name"`
	Kind   string            `json:"kind"`
	Labels map[string]string `json:"labels"`
}

// gRPC request

use the same name with config
type CreateDeploymentChainRequest struct {
  ID string
  Applications []ChainApplicationMatcher
}

type ChainApplicationMatcher struct {
  ...
}

// model in Proto format

type DeploymentChain struct {
    ID string
    Blocks []ChainBlock
}

type ChainBlock struct {
    Nodes []ChainNode
}

type ChainNode struct {
     Application *ChainApplicationRef
     Deployment *ChainDeploymentRef
     Runnable bool
}

// Just needed information to render on the web.
type ChainApplicationRef struct {
   AppID string
   AppName string
}

// Just needed information to render on the web.
type ChainDeploymentRef struct {
  DeploymentID string
  ...
  CompletedAt int64
}

```

Reply:
`ChainBlock` or `ChainGroup` (Maybe we should use ChainBlock to indicate that we are using Blockchain technology in our project. 🤣 )

Reply:
- 
```golang
// use the same name with config
type CreateDeploymentChainRequest struct {
  ApplicationMatcher {
  }

  ID string
}
```
The CreateDeploymentChainRequest should not contain the applications field, since we will apply the "matcher", find applications, and make sync command for them in control-plane (piped which is not in charge for those applications should not fetch the application info as well so do it in control-plane is better)

- "filter" or "matcher" are both good to me so using "matcher" is fine 🤣 

other things are as same as my image so 🙌

Reply:
> The CreateDeploymentChainRequest should not contain the applications field, since we will apply the "matcher", find applications

I don't get this. That is the matcher list `  Applications []ApplicationMatcher`. Don't we need them to find apps in the control plane?

Reply:
Ah sorry, the field named `Applications` make me get it wrong 🙏  Should be `ApplicationsMatchers`
 🤔

Reply:
```golang
type ChainBlock struct {
    Nodes []ChainNode
    Runnable bool
}

type ChainNode struct {
     Application *ChainApplicationRef
     Deployment *ChainDeploymentRef
     CompletedAt int64
}
```
btw, I feel like this, the `Runnable` used by the previous block to unlock this node, and the current block node update its CompletedAt and once all nodes in a block completed, the next block is marked as runnable

Reply:
Updated! 👍 

``` go

// config

type DeploymentChain struct {
	ApplicationMatchers []ChainApplicationMatcher`json:"applications"`
	// Conditions *ChainTriggerCondition `json:"conditions,omitempty"`
}

type ChainApplicationMatcher struct {
	Name   string            `json:"name"`
	Kind   string            `json:"kind"`
	Labels map[string]string `json:"labels"`
}

// gRPC request

// use the same name with config
type CreateDeploymentChainRequest struct {
  ApplicationMatcher {
  }

  ID string
  ApplicationMatchers []ApplicationMatcher
}

// model in Proto format

type DeploymentChain struct {
    ID string
    Blocks []ChainBlock
}

type ChainBlock struct {
    Nodes []ChainNode
}

type ChainNode struct {
     Application *ChainApplicationRef
     Deployment *ChainDeploymentRef
}

// Just needed information to render on the web.
type ChainApplicationRef struct {
   AppID string
   AppName string
}

// Just needed information to render on the web.
// Basically we just need the "status" and ID to link to the actual one.
type ChainDeploymentRef struct {
  DeploymentID string
  ...
  CompletedAt int64
}

```

Reply:
About `runnable` I am thinking that maybe we don't need it.
Because in case of having that field, who will have the responsibility to set that value?
If a block contains multiple applications which are managed by different Pipeds, so which piped will mark that `runnable field for the next block?

The control plane can run a simple check every time based on the deployment status to decide?
How do you think?

Reply:
> who will have the responsibility to set that value?

control-plane will do it, when pipedx submit its deployment state (as successes), we will check for state of all node in its block, if the block's nodes are all completed, mark the runnable state for next block as true, that's my plan. How do you think?

Reply:
> The control plane can run a simple check every time based on the deployment status to decide?

Yeh, the control-plane run the check, but the pipedx who deploy the deployment just internally check does it should start trigger its deployment, so we need that runnable lock, I think

Reply:
I got your points. Looks nice.

Just a small thing, about where to put runnable field, if that field will be set by control-plne so I still prefer to put it in each node. By that way we still be abe to get what we want and beside that we can have more control on running order of each application. I mean controlling an application gives us more opportunity than controlling whole block.
(example, we allow apps in a block could be run serially.)

Reply:
okay, I get that. For now, it will be lock/unlock all nodes of a block at the same time, right?

Reply:
note: we may don't need that lock, since the control-plane knows when the previous block deployed successfully, one rpc call to ask control-plane whether the pipedx should start its deployment is enough 👍

Reply:
Cool. 👍